File: src/main/java/gargoyle/ct/mutex/SocketMutex.java
package gargoyle.ct.mutex;

import gargoyle.ct.log.Log;

import java.io.IOException;
import java.net.ServerSocket;

public final class SocketMutex implements CTMutex {

    private final int          port;
    private       ServerSocket mutex;

    public SocketMutex(int port) {
        this.port = port;
    }

    @Override
    public synchronized boolean acquire() {
        try {
            mutex = new ServerSocket(port);
            return true;
        } catch (IOException ex) {
            return false;
        }
    }

    @Override
    public synchronized void release() {
        try {
            if (mutex != null) {
                if (!mutex.isClosed()) {
                    mutex.close();
                }
                mutex = null;
            }
        } catch (IOException ex) {
            Log.error(ex, MSG_MUTEX_ERROR);
        }
    }
}

File: util/src/main/kotlin/gargoyle/ct/util/mutex/impl/SocketMutex.kt
package gargoyle.ct.util.mutex.impl

import gargoyle.ct.util.log.Log
import gargoyle.ct.util.mutex.CTMutex
import java.io.IOException
import java.net.ServerSocket

class SocketMutex(private val port: Int) : CTMutex {
    private lateinit var mutex: ServerSocket

    @Synchronized
    override fun acquire(): Boolean =
        try {
            mutex = ServerSocket(port)
            true
        } catch (ex: IOException) {
            false
        }

    @Synchronized
    override fun release() {
        try {
            //                mutex = null
            if (::mutex.isInitialized && !mutex.isClosed) mutex.close()
        } catch (ex: IOException) {
            Log.error(ex, "socket mutex error")
        }
    }

}
